phd-tests/framework/src/test_vm/mod.rs

```diff
@@ -780,35 +776,94 @@ impl TestVm {
+        // server process, won't run until this work is complete.)
+        //
+        // Send the task back to the framework so that the test runner can wait
+        // for all under-destruction VMs to be reaped before exiting.
+        let client = self.client.clone();
+        let server = self.server.take();
+        let span = self.tracing_span.clone();
+        let task = tokio::task::spawn(
+            async move {
+                let _server = server;
+                match client
+                    .instance_get()
+                    .send()
+                    .await
+                    .map(|r| r.instance.state)
+                {
+                    Ok(InstanceState::Destroyed) => return,
+                    Err(e) => warn!(
+                        ?e,
+                        "error getting instance state from dropped VM"
+                    ),
+                    Ok(_) => {}
+                }
+
+                info!("stopping test VM on drop");
+                if let Err(e) = client
+                    .instance_state_put()
+                    .body(InstanceStateRequested::Stop)
+                    .send()
+                    .await
+                {
+                    error!(?e, "error stopping dropping VM");
                     return;
                 }
-                if let Err(err) = self.wait_for_state(
-                    InstanceState::Destroyed,
-                    Duration::from_secs(5),
-                ) {
-                    warn!(?err, "Test VM failed to clean up");
+
+                let check_destroyed = || async {
+                    match client
+                        .instance_get()
+                        .send()
+                        .await
+                        .map(|r| r.instance.state)
+                    {
+                        Ok(InstanceState::Destroyed) => Ok(()),
+                        Ok(state) => {
+                            Err(backoff::Error::transient(anyhow::anyhow!(
+                                "instance not destroyed yet (state: {:?})",
+                                state
+                            )))
+                        }
+                        Err(e) => {
+                            error!(%e, "failed to get state of dropping VM");
+                            Err(backoff::Error::permanent(e.into()))
+                        }
+                    }
+                };
+
+                if backoff::future::retry(
+                    backoff::ExponentialBackoff {
+                        max_elapsed_time: Some(std::time::Duration::from_secs(
+                            5,
+                        )),
+                        ..Default::default()
+                    },
+                    check_destroyed,
+                )
+                .await
+                .is_err()
+                {
+                    error!("dropped VM not destroyed after 5 seconds");
```

Comment:
style nit: i, personally, don't love `if`s with multi-line expressions in the condition. what do you think of
```suggestion
                let destroyed = backoff::future::retry(
                      backoff::ExponentialBackoff {
                          max_elapsed_time: Some(std::time::Duration::from_secs(
                              5,
                          )),
                          ..Default::default()
                      },
                      check_destroyed,
                  )
                  .await;
                if let Some(error) = destroyed {
                    error!(%error, "dropped VM not destroyed after 5 seconds");
```